api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -616,20 +616,26 @@ public User saveUserProperties(Map<String, String> properties) {
 	@Override
 	@Authorized(PrivilegeConstants.EDIT_USER_PASSWORDS)
 	@Logging(ignoredArgumentIndexes = { 1, 2 })
-	public void changePassword(User user, String oldPassword, String newPassword) throws APIException {
+	public void changePassword(User user, String secret, String newPassword) throws APIException {
 		if (user.getUserId() == null) {
 			throw new APIException("user.must.exist", (Object[]) null);
 		}
-		if (oldPassword == null) {
+		if (secret == null) {
 			if (!Context.hasPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS)) {
 				throw new APIException("null.old.password.privilege.required", (Object[]) null);
 			}
-		} else if (!dao.getLoginCredential(user).checkPassword(oldPassword)) {
-			throw new APIException("old.password.not.correct", (Object[]) null);
+		} else if (!dao.getLoginCredential(user).checkPassword(secret) && 
+				!isSecretAnswer(user, secret)) {
+			throw new APIException("old.password.or.secret.not.correct", (Object[]) null);
 		}
 	
 		OpenmrsUtil.validatePassword(user.getUsername(), newPassword, user.getSystemId());
 		dao.changePassword(user, newPassword);
 	}
+
+	@Override
+	public void changePassword(User user, String newPassword) throws APIException {
+		changePassword(user, null, newPassword);
```

Comment:
What about the business rule in `changePassword(User, String, String)` that checks if user has required privilege. See [here](https://github.com/geoff-wasilwa/openmrs-core/blob/TRUNK-3871/api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java#L623)